[effects] Add way to register effectul op

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122348
* #122347

This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.

angelayi committed:
[effects] Add way to register effectul op

[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -199,6 +213,130 @@ def f(x):
         res.sum().backward()
 
+    def test_register_effectful_custom_op(self):
+        lib = torch.library.Library("mylib", "FRAGMENT")  # noqa: TOR901
+        try:
```
use torch.library._scoped_library

angelayi replied:
I just copied it from [other test cases](https://www.internalfb.com/code/search?q=regex%3Aon%20repo%3Afbcode%20torch.library.Library.*%23%20noqa%3A%20TOR901) 😅 but now I know!

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -199,6 +213,130 @@ def f(x):
+
+            # Pytorch custorm op implementation
+            @torch.library.impl(
+                "mylib::record_scalar_tensor", "CompositeExplicitAutograd"
```
need to pass lib=lib

zou3519 commented on test/higher_order_ops/test_with_effects.py:
```diff
@@ -199,6 +213,130 @@ def f(x):
+                return
+
+            # Meta function of the custom op
+            @torch.library.impl_abstract("mylib::record_scalar_tensor")
```
lib=lib

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi commented:
@pytorchbot merge

angelayi committed:
Update on "[effects] Add way to register effectul op"


This adds a way to register an operator as being effectful. I also added a test case which mimics our solution for intermediate logging ([doc](https://docs.google.com/document/d/1eLyGDVe4iplVFiO0I021uLgA4Y6HxK9eqn55e9KzQkc/edit#heading=h.uwec2ukkwhea)), which is by creating a custom op and registering it as effectful.


[ghstack-poisoned]

angelayi commented:
@pytorchbot merge
